### PR TITLE
Introduce block authorship soft deadline

### PR DESCRIPTION
**Pull request by tomusdrw**
Currently, when packing transactions to a block we have few conditions when we decide the block is ready to be signed by the consensus engine and gossiped to the network:
1. block size limit is reached,
2. block weight limit is reached (indirectly through "exhausts resources" errors)
3. we run out of transactions in the transaction pool,
4. deadline to produce the block is reached.

The first 3 I'd consider "regular" conditions, the last one is rather a safety valve, ensuring that the block producer does not miss it's slot due to some irregular slowness.

However, since optimal block packing is a difficult problem (i.e. producing a block with optimal utilisation, while maximising the cumulative `priority` of all transactions in that block), we use a very simple greedy heurstic currently, with a twist, that whenever we run into first transaction that reports resources exhaustion, we attempt to insert at least `MAX_SKIPPED_TRANSACTIONS` more to the block before concluding it's actually full.

This heuristic is obviously sub-optimal and can be gamed, hence the PR introduces another variant of the heuristic to potentially increase block utilisation, but without reaching the hard `deadline` defined by the consensus engine.

The PR introduces "soft deadline" (half of the hard deadline time). Before soft deadline is reached we can try as many transactions as desired from the transaction pool, even if they report resources exhaustion. After soft deadline we switch to the previous heuristic of trying at most `MAX_SKIPPED_TRANSACTIONS`.

### CODE REVIEW

**Comment by gilescope** [comment]
What's the smallest size a tx can be? Maybe we should be checking if block_size + MIN_POSSIBLE_TX_SIZE > block_size_limit because there's always going to be a little bit left over but there may be zero chance that any tx can realistically fit in there.

**Comment by tomusdrw** [comment]
> What's the smallest size a tx can be?

That's a very good point, but it's up to the runtime to decide and we currently have no good way to ask the runtime about that. Also note that any additional heuristic we add, it should be added for both the size limit and the "resource/weight" limit to make sure we cater for all kinds of runtimes.

**Comment by tomusdrw** [comment]
@bkchr can you take a look if soft deadline is okay, or if you'd rather prefer to add transactions up to the hard deadline?

**Comment by gww-parity** [comment]
What about failed test:

```
failures:
---- basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping stdout ----
thread 'basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `2`', client/basic-authorship/src/basic_authorship.rs:759:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:92:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
   4: sc_basic_authorship::basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping::{{closure}}
   5: sc_basic_authorship::basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping
   6: core::ops::function::FnOnce::call_once
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
failures:
    basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping
test result: FAILED. 6 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.25s
error: test failed, to rerun pass '-p sc-basic-authorship --lib'
```

?

File: client/basic-authorship/src/basic_authorship.rs [hunk 1]
```diff
@@ -845,4 +875,141 @@ mod tests {
...
+			Box::new(move || {
+				let mut value = cell.lock();
+				let old = *value;
+				*value = old + time::Duration::from_secs(1);
+				old
+			}),
+		);
+
+		// when
+		let deadline = time::Duration::from_secs(300);
```
Comment (gww-parity) [inline client/basic-authorship/src/basic_authorship.rs:932]: as this number is hardcoded:

* should it be explained in few words why 300 and when may user want to tweak it?
* or maybe make const out of it?

**Review by gww-parity — COMMENTED** [review]
what's below 
+ https://github.com/paritytech/substrate/pull/9663#issuecomment-929325767 of test

File: client/basic-authorship/src/basic_authorship.rs (hunk 1 above)
Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:932]: This is a test.

File: client/basic-authorship/src/basic_authorship.rs [hunk 2]
```diff
@@ -386,6 +390,13 @@ where
 						MAX_SKIPPED_TRANSACTIONS - skipped,
 					);
 					continue
+				} else if (self.now)() < soft_deadline {
```
Comment (bkchr) [inline client/basic-authorship/src/basic_authorship.rs:393]:
```suggestion
				} else if now < soft_deadline {
```

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
> What about failed test:
> 
> ```
> failures:
> ---- basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping stdout ----
> thread 'basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping' panicked at 'assertion failed: `(left == right)`
>   left: `1`,
>  right: `2`', client/basic-authorship/src/basic_authorship.rs:759:13
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:515:5
>    1: core::panicking::panic_fmt
>              at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/panicking.rs:92:14
>    2: core::panicking::assert_failed_inner
>    3: core::panicking::assert_failed
>    4: sc_basic_authorship::basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping::{{closure}}
>    5: sc_basic_authorship::basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping
>    6: core::ops::function::FnOnce::call_once
>              at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/core/src/ops/function.rs:227:5
> note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> failures:
>     basic_authorship::tests::should_not_remove_invalid_transactions_when_skipping
> test result: FAILED. 6 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.25s
> error: test failed, to rerun pass '-p sc-basic-authorship --lib'
> ```
> 
> ?

This is a flaky test

**Review by gww-parity — APPROVED** [review]

File: client/basic-authorship/src/basic_authorship.rs (hunk 1 above)
Comment (gww-parity) [reply client/basic-authorship/src/basic_authorship.rs:932]: Still I don't know why this value for test. To me tests are documentation, and I don't know how to read it -> is it random? Or semi-random? etc. Overall, yes, it's not blocking from approval, just something imho nice to polish.

File: client/basic-authorship/src/basic_authorship.rs (hunk 2 above)
Comment (gww-parity) [reply client/basic-authorship/src/basic_authorship.rs:393]: I wonder, isn't actually original  `(self.now)()` intentional to get more recent reading?

File: client/basic-authorship/src/basic_authorship.rs (hunk 1 above)
Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:932]: Seeing tests as docs is really a bad idea.

Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:932]: https://github.com/paritytech/substrate/blob/9d058d4275955a86dac23c98fbae847c934504f9/primitives/consensus/common/src/lib.rs#L208 here are the docs

Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:932]: The deadline is that high to not have it triggered.

Comment (gww-parity) [reply client/basic-authorship/src/basic_authorship.rs:932]: > Seeing tests as docs is really a bad idea.

I see we come from different background and experience ;).

Otherwise, thank you for pointers and links.

Again, I agree with you that this is just a detail, so whatever will be fine with me (may be even ignored) and as you see my approval for PR was already given regardless :).

Comment (bkchr) [reply client/basic-authorship/src/basic_authorship.rs:932]: > I see we come from different background and experience ;).

I mean, I would not say that I do not read tests from time to time, but this normally means that the docs are shit and I need some examples on how to use it :D

Comment (gww-parity) [reply client/basic-authorship/src/basic_authorship.rs:932]: :). I consider that good tests have a lot of value. They can document expectations in non-ambiguous way (contrary to human natural language), and keep assumptions in check (e.g. setting time limits with not too much slack, can help you in detecting if someone accidentally introduces performance regression (that may be not caught by benchmarking or make diagnosis easier).... maybe not in this particular piece of code but in general). 
Ideally, tests may cover expected and edge cases, so I can read from them who edge cases may look like, but yeah, not always worth the effort.

Regarding code review process, asking questions about tests I consider practice motivating to rethink hidden assumptions, and if we can do better. E.g. here, maybe 300 is 10x too much and maybe we can go 10x lower? Maybe it will help to catch other issues? Or maybe it's not worth it and we keep 300 and move on. :)

TL;DR- to me documentation and tests have potential (but not always have to) document different things, from different angles :).

File: client/basic-authorship/src/basic_authorship.rs (hunk 2 above)
Comment (tomusdrw) [reply client/basic-authorship/src/basic_authorship.rs]: It's a good suggestion, there is nothing intensive happening between these calls, and it's worth saving this one extra syscall. That was my intention when I introduced `now` variable initially as well, but somehow lost it in all the refactorings.

**Comment by tomusdrw** [comment]
bot merge

**Comment by tomusdrw** [comment]
I see, making `soft_deadline(_percentage)` configurable makes sense then.
